### PR TITLE
[feature]allow to attach part between different disks

### PR DESCRIPTION
**Pull request by lingo-xp**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
allow to attach part between different disks or storages

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by lingo-xp** [comment]
#34436

**Comment by lingo-xp** [comment]
when attacing a part from different disk,  this feature will randomly pick a disk from current storage policy, then copy the content of source part to a temp part folder, finally load it  as formal part

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 1]
```diff
@@ -1809,6 +1809,23 @@ void IMergeTreeDataPart::makeCloneOnDisk(const DiskPtr & disk, const String & di
...
 
+void IMergeTreeDataPart::makeCloneToOtherDisk(const DiskPtr & disk, const String & directory_name,const String & relative_data_path) const
+{
+    if (disk->getName() == volume->getDisk()->getName())
+        throw Exception("Can not clone data part " + name + " to same disk " + volume->getDisk()->getName(), ErrorCodes::LOGICAL_ERROR);
+    String path_to_clone = fs::path(relative_data_path) / directory_name;
+    if (disk->exists(path_to_clone))
+    {
+        LOG_WARNING(storage.log, "Path {} already exists. Will remove it and clone again.", fullPath(disk, path_to_clone));
+        disk->removeRecursive(fs::path(path_to_clone) / "");
```
Comment (vdimir) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:1820]: Are we sure that we don't delete any significant data? Maybe let's throw an exception in this case and let user to clean files manually.

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 2]
```diff
@@ -5815,25 +5815,24 @@ MergeTreeData::MutableDataPartPtr MergeTreeData::cloneAndLoadDataPartOnSameDisk(
...
+            is_on_same_disk = true;
             break;
         }
     }
-    if (!does_storage_policy_allow_same_disk)
-        throw Exception(
-            "Could not clone and load part " + quoteString(src_part->getFullPath()) + " because disk does not belong to storage policy",
-            ErrorCodes::BAD_ARGUMENTS);
+
+    auto target_disk = is_on_same_disk ? src_part->volume->getDisk() : getStoragePolicy()->getRandomDisk();
```
Comment (vdimir) [inline src/Storages/MergeTree/MergeTreeData.cpp:5828]: Why do you choose to get random disk? What about taking first one to have expected behavior?

File: tests/config/install.sh [hunk 3]
```diff
@@ -44,6 +44,7 @@ ln -sf $SRC_PATH/config.d/logger.xml $DEST_SERVER_PATH/config.d/
 ln -sf $SRC_PATH/config.d/named_collection.xml $DEST_SERVER_PATH/config.d/
 ln -sf $SRC_PATH/config.d/ssl_certs.xml $DEST_SERVER_PATH/config.d/
 ln -sf $SRC_PATH/config.d/filesystem_cache_log.xml $DEST_SERVER_PATH/config.d/
+ln -sf $SRC_PATH/config.d/local_disks.xml $DEST_SERVER_PATH/config.d/
```
Comment (vdimir) [inline tests/config/install.sh:47]: If we need this only for one test it's better to make it integration test with custom config

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 2 above)
Comment (lingo-xp) [reply src/Storages/MergeTree/MergeTreeData.cpp:5828]: I used getAnyDisk() at first time,  but in our cluster, we have mutiple hdd disks in a single machine, in order to improve total IO performance, we try to make data more even, so I choose random strategy here. maybe we can put a setting  here, let user to decide which strategy they want.

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp (hunk 1 above)
Comment (lingo-xp) [reply src/Storages/MergeTree/IMergeTreeDataPart.cpp:1820]: directory_name is a temp part name, path_to_clone is the path of directory_name,  besides, we have already checked the exsitence of  formal part path  in cloneAndLoadDataPart function,  so I think we can remove this temp folder here.

File: tests/config/install.sh (hunk 3 above)
Comment (lingo-xp) [reply tests/config/install.sh:47]: Sorry, I am not so familiar with current test workflow, is there any example to define a storage policy in custum config

Comment (vdimir) [reply tests/config/install.sh:47]: For example https://github.com/ClickHouse/ClickHouse/tree/master/tests/integration/test_multiple_disks

File: src/Storages/MergeTree/MergeTreeData.cpp (hunk 2 above)
Comment (vdimir) [reply src/Storages/MergeTree/MergeTreeData.cpp:5828]: Yes, setting is ok, I suppose

**Comment by lingo-xp** [comment]
@vdimir please review the latest code.

**Comment by vdimir** [comment]
> test_multiple_disks/test.py::test_move_across_policies_does_not_work | FAIL

https://s3.amazonaws.com/clickhouse-test-reports/37478/954f4dd87cf95a2f898859402fa57b21f8f42ff5/integration_tests__release__actions__[2/2].html

It expect that such `ATTACH` would not work, so we can just fix the test

> AST fuzzer (UBSan, actions) — Logical error: 'Function lessOrEquals with argument with type Nothing and default

Known issue https://github.com/ClickHouse/ClickHouse/issues/37985

**Comment by vdimir** [comment]
@lingo-xp 

I've made some fixes (style and test) in https://github.com/ClickHouse/ClickHouse/tree/37478-followup you can fetch commits (I can't push directly to PR because access is not enabled)

**Review by vdimir — APPROVED** [review]

**Comment by lingo-xp** [comment]
@vdimir  
Thank you! I have already merged your commits, we can move to test step again

**Pull request closed without merging**